Integrate the Google Play Billing Library into engine

qiuguohua opened the pull request:
Re: #
https://developer.android.com/google/play/billing/integrate
https://developer.android.com/reference/com/android/billingclient/api/BillingClient?_gl=1*11tcdon*_up*MQ..*_ga*MjcyNzYxMzIuMTczMTM3NTA0MQ..*_ga_6HH9YJMN9M*MTczMTM3NTA0MS4xLjAuMTczMTM3NTA0MS4wLjAuNzA1NTQwMDMy
### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!
- [ ] Your pull request title is using English, it's precise and appropriate.
- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.
- [ ] Document new code with comments in source code based on API docs
- [ ] Make sure any runtime log information in `log` , `error` or `new Error('')` has been moved into `EngineErrorMap.md` with an ID, and use `logID(id)` or `new Error(getError(id))` instead.
- To official teams:
  - [ ] Check that your PR is following our [guides](https://github.com/cocos/3d-tasks/blob/master/workflows/readme.md)
-->

bofeng-song committed:
Fix the issue where the return value of getSafeArea in WeChat may be undefined. (#17831)

dumganhar committed:
[v3.8.5] Fix wrong head commit id in web-interface-check.yml ci script. (#17827)

Commit:
Integrate the Google Play Billing Library into engine

sablib committed:
fix TypedArray set (#17817)

* fix TypedArray set

* Fix TypedArrayTemp::set function templates

qiuguohua commented on cocos/core/platform/billing/billing.ts:
```diff
@@ -0,0 +1,804 @@
+     * 这个字段已经废弃。
+     * 看看SERVICE_UNAVAILABLE将使用哪一个来代替此代码。
+     */
+    SERVICE_TIMEOUT = -3,
```
This field is marked as deprecated in the google docs, does it need to be removed?

minggo replied:
Yep, i think should remove it. But it is better to listen to the google guys' opinions too.

qiuguohua commented on cocos/core/platform/billing/billing.ts:
```diff
@@ -0,0 +1,804 @@
+export class GooglePlayBilling {
+    private _eventTarget: EventTarget = new EventTarget();
+    constructor () {
+        if (!JSB || !jsb.billing) {
```
Return if non-JSB or jsb.billing is null, only in android platform jsb.billing is not null

minggo replied:
Should only work on google play build target.

Commit:
1. Remove redundant code
2. Adjust the file name

GengineJS committed:
Fix surface-shader (#17837)

qiuguohua commented on cocos/native-binding/index.ts:
```diff
@@ -1476,4 +1476,349 @@ export declare namespace native {
+         */
+        readonly signature: string;
+
+        readonly toStr: string;
```
The name should be toString, but ts has an internal toString, so it's changed to toStr.

qiuguohua commented on native/cocos/platform/android/modules/google_play/billing/JniBillingHelper.cpp:
```diff
@@ -0,0 +1,583 @@
+    jintArray result = env->NewIntArray(size);
+    jint* buf = new jint[size];
+    for (int i = 0; i < size; ++i) {
+        buf[i] = productDetailsList[i]->hashCode;
```
Finding objects in Java using hashCode

qiuguohua commented:
@cocos-robot run test cases

Commit:
Optimized code

dumganhar commented on cocos/native-binding/index.ts:
```diff
@@ -1476,4 +1476,350 @@ export declare namespace native {
+         * @en Represents the available purchase plans to buy a subscription product.
+         * @zh 代表一次性购买产品的报价详情。
+         */
+        readonly hashCode: number;
```
Is the comment for `hashCode` wrong?

dumganhar commented on cocos/native-binding/index.ts:
```diff
@@ -1476,4 +1476,350 @@ export declare namespace native {
+         */
+        readonly title: string;
+
+        readonly toStr: string;
```
Missing API comment for `toStr`

dumganhar commented on native/cocos/bindings/manual/jsb_billing.cpp:
```diff
@@ -0,0 +1,1461 @@
+
+bool jsb_register_all_billing(se::Object* obj) { // NOLINT
+    se::Value nsVal;
+    if (!obj->getProperty("jsb", &nsVal, true)) {
```
Should use another namespace, like `gp` or `googleplay`

knoxHuang committed:
refine code (#17824)

dumganhar committed:
[v3.8.5] Remove unused exports/wait-for-ammon-instantiation.ts (#17834)

dumganhar committed:
[v3.8.5] Add default constructor for classes (#17746)

longchuan committed:
fix the bug of bt._safe_delete (#17842)

dumganhar committed:
[v3.8.5] Remove unused baidu platform adapter. (#17845)

* [v3.8.5] Remove unused baidu platform adapter.

* Update minigame/README.md

dumganhar committed:
[v3.8.5] Don't use private property '_files' in cache-manager.js. (#17846)

dumganhar committed:
fixed #17833: [v3.8.5] Add an ability to load wasm/asmjs module manually (#17844)

dogodo-cc committed:
feat: add flags config (#17848)

* feat: add flags config

* feat: update doc

cocos-robot committed:
[ci skip][AUTO]: Automated code generating update: bdf2e40bdfe04b407d943a6be863166b68446b06  (#17848) (#17851)

Co-authored-by: cocos-robot <cocos-robot@cocos.com>

GengineJS committed:
Fixed a bug with dof under vulkan back end (#17847)

cocos-robot committed:
[ci skip][AUTO]: Automated code generating update: c48028006ba765027fe62200c17b4fce5c7b86c5  (#17844) (#17850)

Co-authored-by: cocos-robot <cocos-robot@cocos.com>

W-MTZing committed:
[v3.8.5] Remove unused cocos play，link-sure, qtt  platform adapter (#17849)

Co-authored-by: lbq <563294062@qq.com>

Commit:
优化

star-e committed:
refactored builtin-pipeline (#17860)

GengineJS committed:
[v3.8.5] Fix Model.setInstancedAttribute doesn't support attribute of INT format. (#17858)

* setInstancedAttribute

* Add copyTypedArray

* Fix Model::setInstancedAttribute and SubModule::setInstancedAttribute.

---------

Co-authored-by: James Chen <dumganhar@gmail.com>

dumganhar committed:
[v3.8.5] Mangle private properties in touch-manager.ts (#17868)

* [v3.8.5] Mangle private properties in touch-manager.ts

* Fix eslint errors in mouse-input.ts

knoxHuang committed:
inspector adds prefab and spine previews (#17855)

dumganhar committed:
[v3.8.5] Make dispatchImmediately variable be a const value which will help treeshaking. (#17853)

star-e committed:
fix camera id (#17857)

knoxHuang committed:
Fix editor preview, canvas component calculates Camera invalidly (#17856)

dumganhar committed:
Update ccbuild to 2.2.19 (#17870)

star-e committed:
fix native viewport (#17871)

dogodo-cc committed:
feat: add physics-2d-box2d-wasm options (#17874)

star-e committed:
fix legacy pipeline viewport (#17875)

dumganhar committed:
fixed #17867: [bug] b2RigidBody2D was not removed from b2PhysicsWorld _animatedBodies container if its type was changed. (#17873)

qiuguohua committed:
V3.8.5 oh api12 (#17869)

* Improvement of openharmony's editbox (#17042)

* Improvement of openharmony's editbox

* Remove redundant header file

---------

Co-authored-by: qiuguohua <guohua.qiu@cocos.com>
# Conflicts:
#	templates/openharmony/entry/src/main/cpp/types/libcocos/index.d.ts
#	templates/openharmony/entry/src/main/ets/components/EditBoxDialog.ets
#	templates/openharmony/entry/src/main/ets/pages/index.ets
#	templates/openharmony/entry/src/main/ets/workers/cocos_worker.ts

* Upgrade the openharmony API to 12

* Remove debugging information

* Fix some bugs in the ark engine
Fix some issues with jsvm
* Release module_info

* Fix std::any_of return value not being used

---------

Co-authored-by: qiuguohua <guohua.qiu@cocos.com>

dumganhar committed:
[v3.8.5] Update CI for HarmonyOS Next. (#17879)

dumganhar committed:
[ci skip] Remove unused code in generate-oh-sdk-cache.yml (#17881)

Commit:
1. increase the platform template of google play
2. add vendor module
3. Fix some problems of google play.
